axum/src/typed_header.rs

```diff
@@ -61,17 +61,19 @@ where
     type Rejection = TypedHeaderRejection;
 
     async fn from_request_parts(parts: &mut Parts, _state: &S) -> Result<Self, Self::Rejection> {
-        match parts.headers.typed_try_get::<T>() {
-            Ok(Some(value)) => Ok(Self(value)),
-            Ok(None) => Err(TypedHeaderRejection {
+        let mut values = parts.headers.get_all(T::name()).iter();
+        let is_missing = values.size_hint() == (0, Some(0));
+        T::decode(&mut values)
+            .map(Self)
+            .map_err(|err| TypedHeaderRejection {
                 name: T::name(),
-                reason: TypedHeaderRejectionReason::Missing,
-            }),
-            Err(err) => Err(TypedHeaderRejection {
-                name: T::name(),
-                reason: TypedHeaderRejectionReason::Error(err),
-            }),
-        }
+                reason: if is_missing {
+                    // Report a more precise rejection for the missing header case.
+                    TypedHeaderRejectionReason::Missing
```

Comment:
I'd be fine removing this variant, personally.

Reply:
Yeah I was a bit torn.

The downsides of doing so are:

1. Slightly worse error message / debugging experience.
2. That kind of change would certainly be breaking as users might be checking directly against the error enum.

It seems fairly low-cost to keep it for backward compat and for the slightly better error message for now.

Reply:
I think we should keep it. For example for auth headers it feels valuable to be able to distinguish a missing header from an invalid header value. Stuff like https://github.com/tokio-rs/axum/issues/316 comes to mind.